### PR TITLE
hwdb: Acer Iconia Tab W500 support

### PR DESCRIPTION
**Pull request by yaspoon**
The Acer W500 has a Bosch BMA150 accelerometer which
is mounted in a different orientation to the display
orientation.

Correct the accelerometer output to match the screen
orientation.

### CODE REVIEW

**Comment by yaspoon** [comment]
I posted this patch as a question on the mailing list and was asked to instead add it as a PR.
The problem is that I can't seem to get the modalias string to be generic enough like the other strings present in the file for the Asus and Winbook devices.

I'm guessing that the current string in the patch will match any BMA150 device not just the one found in the Acer Iconia Tab W500. So I'm wondering how I can generate a more accurate device matching string?

I was also told to CC in @hadess

**Comment by yaspoon** [comment]
Something I forgot to add is the udevadm output.
Also the device doesn't show up as a /dev/iio:device*.
It shows up as an input as you can see in the udevadm output
is this a potential issue?

```
udevadm info --path=/devices/virtual/input/input17
P: /devices/virtual/input/input17
E: ABS=7
E: DEVPATH=/devices/virtual/input/input17
E: EV=9
E: ID_INPUT=1
E: ID_INPUT_ACCELEROMETER=1
E: ID_INPUT_ACCELEROMETER_ORIENTATION=right-up
E: IIO_SENSOR_PROXY_TYPE=input-accel
E: MODALIAS=input:b0019v0000p0000e0000-e0,3,kra0,1,2,mlsfw
E: NAME="Acer BMA150 accelerometer"
E: PHYS="wmi/input1"
E: PRODUCT=19/0/0/0
E: PROP=0
E: SUBSYSTEM=input
E: SYSTEMD_WANTS=iio-sensor-proxy.service
E: TAGS=:seat:systemd:
E: USEC_INITIALIZED=121291851
```

**Comment by poettering** [comment]
@hadess, say something!

File: hwdb/60-sensor.hwdb [hunk 1]
```diff
@@ -42,6 +42,12 @@
 # Sort by brand, model
 
 #########################################
+# Acer
+#########################################
+evdev:input:b0019v0000p0000e0000*
```
Comment (hadess) [inline hwdb/60-sensor.hwdb]: That's not the correct syntax for this file, see the comments at the top of the file for the correct syntax.

**Comment by yaspoon** [comment]
I've added the sensor:modalias: to the beginning of the match string. But this now no longer matches and the accelerometer isn't corrected. I also tried sensor:modalias:input:b0019v0000p0000e0000* 
and that didn't match either. I don't understand what the <parent device modalias> means and how to find it for my device. The dmi:<dmi string> part I can probably figure out using dmidecode but it won't even match on the sensor let along trying to match it on the tablet model too.

**Comment by hadess** [comment]
> I've added the sensor:modalias: to the beginning of the match string. But this now no longer matches and the accelerometer isn't corrected. I also tried sensor:modalias:input:b0019v0000p0000e0000*
and that didn't match either. I don't understand what the means and how to find it for my device. The dmi: part I can probably figure out using dmidecode but it won't even match on the sensor let along trying to match it on the tablet model too.

What does, "it doesn't match" mean? What are you running to see whether it matches? And finally, which version of systemd are you using?

**Comment by yaspoon** [comment]
When I say "it doesn't match" what I mean is that when I add the "sensor:modalias:" to the front of the match string making the string "sensor:modalias:evdev:input:b0019v0000p0000e0000*" the accelerometer orientation is not corrected. So really I'm just guessing that it's not working. The way I test it is to change the match string to have "sensor:modalias:" at the front then I run `systemd-hwdb update` and I reboot because I can't get the "udevadm trigger" command to work because my accelerometer does not show up as a /dev/iio:device*. When I then log in the screen will be rotated 90 degrees to the left of where is should be because systemd/udev doesn't seem to be correcting the orientation. Is there a better way to test this?

The software versions I am using are below.
ArchLinux
systemd:232-7
Linux 4.8.13-1
iio-sensor-proxy:2.0-1

**Comment by hadess** [comment]
> systemd:232-7

Does that include the patches to add the sensor hwdb to systemd? Because they weren't in a release yet.

**Comment by yaspoon** [comment]
I wasn't sure whether the Arch version did have the patches you mentioned pretty sure it doesn't.... So I've now built and installed systemd from master. I'm still having the same problem though of the accelerometer orientation not being corrected when using the current patches in the PR.

After reading the 60-sensor.hwdb file again I realised I hadn't posted `udevadm info --export-db` so that should now be attached to this comment.

The file also mentions I should give a description of the device but I don't know a huge amount about the device to give a description of it. Other then what is given in the the udev output. Its a "Acer BMA150 Accelerometer" it shows up as an input not as a /dev/iio:device* (The light sensor does show up as a /dev/iio:device* which is interesting).
It's actually a Bosch BMA150 accelerometer and the datasheet is available here:http://odroid.com/dokuwiki/lib/exe/fetch.php?media=en:universal_motion_joypad:bma150.pdf

[udev-db.txt](https://github.com/systemd/systemd/files/770381/udev-db.txt)

**Comment by hadess** [comment]
> It's actually a Bosch BMA150 accelerometer and the datasheet is available 

It isn't, really. It might have been in the original Acer laptop, but it's really a platform driver (implemented in drivers/platform/x86/acer-wmi.c)

> ID_INPUT_ACCELEROMETER_ORIENTATION=...

In the udevadm means that you either captured it with an old systemd, or you didn't update systemd correctly (it was removed in systemd 221).

**Comment by yaspoon** [comment]
That was good spotting. I had forgotten that my first attempt to get auto rotation working was to download both the accelerometer and the udev rule that went with it and to then copy them into the right spots in the filesystem. Then I came accross your iio-sensor-proxy project.

 I've now removed both of those files and tested again but the rotation is still not corrected.

I have captured another udevadm export now that I've removed the files.
[udev-db.txt](https://github.com/systemd/systemd/files/785683/udev-db.txt)

**Comment by hadess** [comment]
https://github.com/systemd/systemd/pull/5510 added supported for platform drivers matching. See the PR there to update your patch. Please make sure to test with systemd 233.

**Comment by poettering** [comment]
@hadess does this look right now?

(the PR is a bit in disarray due to multiple updates. But the effective patch looks OK to me, and could be applied with "squash and merge" removing the multiple individual commits.)

**Comment by hadess** [comment]
I'd still like to see that the quirks are getting applied correctly on the machine before merging anything.

**Comment by keszybz** [comment]
@yaspoon could you confirm that the latest version works correctly?

**Comment by yaspoon** [comment]
I cannot test at the moment because I am getting "Operation not permitted" on the accelerometer device when trying to use iio-sensor-proxy which is apparently due to some changes to the acer_wmi driver. Even though I'm running the 4.11 kernel where the fix has apparently been merged.

**Comment by yaspoon** [comment]
@hadess I downgraded the kernel to 4.10 to get the accelerometer working again. Trying to use #5510 doesn't seem to work. Which I'm guessing is because my device doesn't show up as a platform device. I've uploaded the output of `udevadm info --export-db` here https://gist.github.com/yaspoon/ca55e00a66c3deede2f94a6b7847374f . In #5510 the device is "/devices/platform/lis3lv02d/input/input13" where as the device in my Acer W500 is /devices/virtual/input/input18.

**Pull request closed without merging**